File: system/multitier/backend-typescript/src/core/services/order.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { ILike, Repository } from 'typeorm';
import { randomUUID } from 'node:crypto';
import Decimal from 'decimal.js';
import { Order } from '../entities/order.entity';
import { OrderStatus } from '../entities/order-status.enum';
import { PlaceOrderRequest } from '../dtos/place-order-request.dto';
import { PlaceOrderResponse } from '../dtos/place-order-response.dto';
import {
  BrowseOrderHistoryResponse,
  BrowseOrderHistoryItemResponse,
} from '../dtos/browse-order-history-response.dto';
import { ViewOrderDetailsResponse } from '../dtos/view-order-details-response.dto';
import { ValidationException } from '../exceptions/validation.exception';
import { NotExistValidationException } from '../exceptions/not-exist-validation.exception';
import { ErpGateway } from './external/erp.gateway';
import { ClockGateway } from './external/clock.gateway';
import { TaxGateway } from './external/tax.gateway';
import { CouponService } from './coupon.service';

@Injectable()
export class OrderService {
  private static readonly RESTRICTED_MONTH = 11; // December (0-indexed)
  private static readonly RESTRICTED_DAY = 31;
  private static readonly RESTRICTED_HOUR = 23;
  private static readonly RESTRICTED_MINUTE = 59;

  constructor(
    @InjectRepository(Order)
    private readonly orderRepository: Repository<Order>,
    private readonly erpGateway: ErpGateway,
    private readonly clockGateway: ClockGateway,
    private readonly taxGateway: TaxGateway,
    private readonly couponService: CouponService,
  ) {}

  async placeOrder(request: PlaceOrderRequest): Promise<PlaceOrderResponse> {
    const sku = request.sku;
    const quantity = request.quantity;
    const country = request.country;
    const couponCode = request.couponCode;

    const orderTimestamp = await this.clockGateway.getCurrentTime();

    const utcMonth = orderTimestamp.getUTCMonth();
    const utcDay = orderTimestamp.getUTCDate();

    if (
      utcMonth === OrderService.RESTRICTED_MONTH &&
      utcDay === OrderService.RESTRICTED_DAY
    ) {
      const utcHour = orderTimestamp.getUTCHours();
      const utcMinute = orderTimestamp.getUTCMinutes();

      if (
        utcHour > OrderService.RESTRICTED_HOUR ||
        (utcHour === OrderService.RESTRICTED_HOUR &&
          utcMinute >= OrderService.RESTRICTED_MINUTE)
      ) {
        throw new ValidationException(
          'Orders cannot be placed between 23:59 and 00:00 on December 31st',
        );
      }
    }

    const unitPrice = await this.getUnitPrice(sku);
    const promotion = await this.erpGateway.getPromotionDetails();
    const promotionFactor = promotion.promotionActive ? promotion.discount : 1;
    const basePrice = new Decimal(unitPrice).mul(quantity).mul(promotionFactor).toNumber();

    const discountRate = await this.couponService.getDiscount(couponCode);
    const discountAmount = new Decimal(basePrice).mul(discountRate).toNumber();
    const subtotalPrice = new Decimal(basePrice).sub(discountAmount).toNumber();

    const taxRate = await this.getTaxRate(country);
    const taxAmount = new Decimal(subtotalPrice).mul(taxRate).toNumber();
    const totalPrice = new Decimal(subtotalPrice).add(taxAmount).toNumber();

    const appliedCouponCode = discountRate > 0 ? couponCode ?? null : null;

    const orderNumber = this.generateOrderNumber();

    const order = new Order();
    order.orderNumber = orderNumber;
    order.orderTimestamp = orderTimestamp;
    order.country = country;
    order.sku = sku;
    order.quantity = quantity;
    order.unitPrice = unitPrice;
    order.basePrice = basePrice;
    order.discountRate = discountRate;
    order.discountAmount = discountAmount;
    order.subtotalPrice = subtotalPrice;
    order.taxRate = taxRate;
    order.taxAmount = taxAmount;
    order.totalPrice = totalPrice;
    order.status = OrderStatus.PLACED;
    order.appliedCouponCode = appliedCouponCode ?? null;

    await this.orderRepository.save(order);

    if (appliedCouponCode) {
      await this.couponService.incrementUsageCount(appliedCouponCode);
    }

    const response = new PlaceOrderResponse();
    response.orderNumber = orderNumber;
    return response;
  }

  private async getUnitPrice(sku: string): Promise<number> {
    const productDetails = await this.erpGateway.getProductDetails(sku);
    if (productDetails === null) {
      throw new ValidationException(
        'sku',
        `Product does not exist for SKU: ${sku}`,
      );
    }

    return Number(productDetails.price);
  }

  private async getTaxRate(country: string): Promise<number> {
    const taxDetails = await this.taxGateway.getTaxDetails(country);
    if (taxDetails === null) {
      throw new ValidationException(
        'country',
        `Country does not exist: ${country}`,
      );
    }

    return Number(taxDetails.taxRate);
  }

  async cancelOrder(orderNumber: string): Promise<void> {
    const now = await this.clockGateway.getCurrentTime();

    const utcMonth = now.getUTCMonth();
    const utcDay = now.getUTCDate();

    if (utcMonth === 11 && utcDay === 31) {
      const blackoutStart = new Date(Date.UTC(now.getUTCFullYear(), 11, 31, 22, 0, 0));
      const blackoutEnd = new Date(Date.UTC(now.getUTCFullYear(), 11, 31, 22, 30, 0));

      if (now >= blackoutStart && now <= blackoutEnd) {
        throw new ValidationException(
          'Order cancellation is not allowed on December 31st between 22:00 and 23:00',
        );
      }
    }

    const order = await this.orderRepository.findOne({
      where: { orderNumber },
    });

    if (!order) {
      throw new NotExistValidationException(
        `Order ${orderNumber} does not exist.`,
      );
    }

    if (order.status === OrderStatus.CANCELLED) {
      throw new ValidationException('Order has already been cancelled');
    }

    order.status = OrderStatus.CANCELLED;
    await this.orderRepository.save(order);
  }

  async browseOrderHistory(
    orderNumberFilter?: string,
  ): Promise<BrowseOrderHistoryResponse> {
    let orders: Order[];

    if (!orderNumberFilter || orderNumberFilter.trim() === '') {
      orders = await this.orderRepository.find({
        order: { orderTimestamp: 'DESC' },
      });
    } else {
      orders = await this.orderRepository.find({
        where: {
          orderNumber: ILike(`%${orderNumberFilter.trim()}%`),
        },
        order: { orderTimestamp: 'DESC' },
      });
    }

    const items = orders.map((order) => {
      const item = new BrowseOrderHistoryItemResponse();
      item.orderNumber = order.orderNumber;
      item.orderTimestamp = new Date(order.orderTimestamp).toISOString();
      item.sku = order.sku;
      item.country = order.country;
      item.quantity = order.quantity;
      item.totalPrice = Number(order.totalPrice);
      item.status = order.status;
      item.appliedCouponCode = order.appliedCouponCode;
      return item;
    });

    const result = new BrowseOrderHistoryResponse();
    result.orders = items;
    return result;
  }

  async getOrder(orderNumber: string): Promise<ViewOrderDetailsResponse> {
    const order = await this.orderRepository.findOne({
      where: { orderNumber },
    });

    if (!order) {
      throw new NotExistValidationException(
        `Order ${orderNumber} does not exist.`,
      );
    }

    const response = new ViewOrderDetailsResponse();
    response.orderNumber = order.orderNumber;
    response.orderTimestamp = new Date(order.orderTimestamp).toISOString();
    response.sku = order.sku;
    response.quantity = order.quantity;
    response.unitPrice = Number(order.unitPrice);
    response.basePrice = Number(order.basePrice);
    response.discountRate = Number(order.discountRate);
    response.discountAmount = Number(order.discountAmount);
    response.subtotalPrice = Number(order.subtotalPrice);
    response.taxRate = Number(order.taxRate);
    response.taxAmount = Number(order.taxAmount);
    response.totalPrice = Number(order.totalPrice);
    response.status = order.status;
    response.country = order.country;
    response.appliedCouponCode = order.appliedCouponCode;
    return response;
  }

  private generateOrderNumber(): string {
    return `ORD-${randomUUID().toUpperCase()}`;
  }
}

File: system/monolith/dotnet/Core/Services/OrderService.cs
using Microsoft.EntityFrameworkCore;
using Optivem.Shop.Monolith.Core.Dtos;
using Optivem.Shop.Monolith.Core.Entities;
using Optivem.Shop.Monolith.Core.Exceptions;
using Optivem.Shop.Monolith.Core.Services.External;
using Optivem.Shop.Monolith.Data;

namespace Optivem.Shop.Monolith.Core.Services;

public class OrderService
{
    private static readonly int RestrictedMonth = 12;
    private static readonly int RestrictedDay = 31;
    private static readonly TimeOnly RestrictedTimeStart = new(23, 59);

    private readonly AppDbContext _dbContext;
    private readonly ErpGateway _erpGateway;
    private readonly TaxGateway _taxGateway;
    private readonly ClockGateway _clockGateway;
    private readonly CouponService _couponService;

    public OrderService(AppDbContext dbContext, ErpGateway erpGateway, TaxGateway taxGateway, ClockGateway clockGateway, CouponService couponService)
    {
        _dbContext = dbContext;
        _erpGateway = erpGateway;
        _taxGateway = taxGateway;
        _clockGateway = clockGateway;
        _couponService = couponService;
    }

    public async Task<PlaceOrderResponse> PlaceOrderAsync(PlaceOrderRequest request)
    {
        var sku = request.Sku!;
        var quantity = request.Quantity!.Value;
        var country = request.Country!;
        var couponCode = request.CouponCode;

        var orderTimestamp = await _clockGateway.GetCurrentTimeAsync();

        var utcTime = orderTimestamp.ToUniversalTime();

        if (utcTime.Month == RestrictedMonth && utcTime.Day == RestrictedDay)
        {
            var currentTime = TimeOnly.FromDateTime(utcTime);
            if (currentTime >= RestrictedTimeStart)
            {
                throw new ValidationException("Orders cannot be placed between 23:59 and 00:00 on December 31st");
            }
        }

        var unitPrice = await GetUnitPriceAsync(sku);
        var promotion = await _erpGateway.GetPromotionDetailsAsync();
        var promotionFactor = promotion.PromotionActive ? promotion.Discount : 1.0m;
        var basePrice = unitPrice * quantity;
        var promotedPrice = basePrice * promotionFactor;

        var discountRate = await _couponService.GetDiscountAsync(couponCode);
        var discountAmount = promotedPrice * discountRate;
        var subtotalPrice = promotedPrice - discountAmount;

        var taxRate = await GetTaxRateAsync(country);
        var taxAmount = subtotalPrice * taxRate;
        var totalPrice = subtotalPrice + taxAmount;

        var appliedCouponCode = discountRate > 0m ? couponCode : null;

        var orderNumber = GenerateOrderNumber();

        var order = new Order
        {
            OrderNumber = orderNumber,
            OrderTimestamp = orderTimestamp,
            Country = country,
            Sku = sku,
            Quantity = quantity,
            UnitPrice = unitPrice,
            BasePrice = basePrice,
            DiscountRate = discountRate,
            DiscountAmount = discountAmount,
            SubtotalPrice = subtotalPrice,
            TaxRate = taxRate,
            TaxAmount = taxAmount,
            TotalPrice = totalPrice,
            Status = OrderStatus.PLACED,
            AppliedCouponCode = appliedCouponCode
        };

        _dbContext.Orders.Add(order);
        await _dbContext.SaveChangesAsync();

        if (appliedCouponCode != null)
        {
            await _couponService.IncrementUsageCountAsync(appliedCouponCode);
        }

        return new PlaceOrderResponse { OrderNumber = orderNumber };
    }

    private async Task<decimal> GetUnitPriceAsync(string sku)
    {
        var productDetails = await _erpGateway.GetProductDetailsAsync(sku);
        if (productDetails == null)
        {
            throw new ValidationException("sku", $"Product does not exist for SKU: {sku}");
        }

        return productDetails.Price;
    }

    private async Task<decimal> GetTaxRateAsync(string country)
    {
        var taxDetails = await _taxGateway.GetTaxDetailsAsync(country);
        if (taxDetails == null)
        {
            throw new ValidationException("country", $"Country does not exist: {country}");
        }

        return taxDetails.TaxRate;
    }

    public async Task DeliverOrderAsync(string orderNumber)
    {
        var order = await _dbContext.Orders
            .FirstOrDefaultAsync(o => o.OrderNumber == orderNumber);

        if (order == null)
        {
            throw new NotExistValidationException($"Order {orderNumber} does not exist.");
        }

        if (order.Status != OrderStatus.PLACED)
        {
            throw new ValidationException("Order cannot be delivered in its current status");
        }

        order.Status = OrderStatus.DELIVERED;
        await _dbContext.SaveChangesAsync();
    }

    public async Task CancelOrderAsync(string orderNumber)
    {
        var now = (await _clockGateway.GetCurrentTimeAsync()).ToUniversalTime();

        if (now.Month == 12 && now.Day == 31)
        {
            var currentTime = TimeOnly.FromDateTime(now);
            var cancelBlackoutStart = new TimeOnly(22, 0);
            var cancelBlackoutEnd = new TimeOnly(22, 30);

            if (currentTime >= cancelBlackoutStart && currentTime <= cancelBlackoutEnd)
            {
                throw new ValidationException("Order cancellation is not allowed on December 31st between 22:00 and 23:00");
            }
        }

        var order = await _dbContext.Orders
            .FirstOrDefaultAsync(o => o.OrderNumber == orderNumber);

        if (order == null)
        {
            throw new NotExistValidationException($"Order {orderNumber} does not exist.");
        }

        if (order.Status == OrderStatus.CANCELLED)
        {
            throw new ValidationException("Order has already been cancelled");
        }

        order.Status = OrderStatus.CANCELLED;
        await _dbContext.SaveChangesAsync();
    }

    public async Task<BrowseOrderHistoryResponse> BrowseOrderHistoryAsync(string? orderNumberFilter)
    {
        List<Order> orders;

        if (string.IsNullOrWhiteSpace(orderNumberFilter))
        {
            orders = await _dbContext.Orders
                .OrderByDescending(o => o.OrderTimestamp)
                .ToListAsync();
        }
        else
        {
            var filter = orderNumberFilter.Trim();
            orders = await _dbContext.Orders
                .Where(o => EF.Functions.ILike(o.OrderNumber, $"%{filter}%"))
                .OrderByDescending(o => o.OrderTimestamp)
                .ToListAsync();
        }

        var items = orders.Select(order => new BrowseOrderHistoryItemResponse
        {
            OrderNumber = order.OrderNumber,
            OrderTimestamp = order.OrderTimestamp,
            Sku = order.Sku,
            Country = order.Country,
            Quantity = order.Quantity,
            TotalPrice = order.TotalPrice,
            Status = order.Status,
            AppliedCouponCode = order.AppliedCouponCode
        }).ToList();

        return new BrowseOrderHistoryResponse { Orders = items };
    }

    public async Task<ViewOrderDetailsResponse> GetOrderAsync(string orderNumber)
    {
        var order = await _dbContext.Orders
            .FirstOrDefaultAsync(o => o.OrderNumber == orderNumber);

        if (order == null)
        {
            throw new NotExistValidationException($"Order {orderNumber} does not exist.");
        }

        return new ViewOrderDetailsResponse
        {
            OrderNumber = order.OrderNumber,
            OrderTimestamp = order.OrderTimestamp,
            Sku = order.Sku,
            Quantity = order.Quantity,
            UnitPrice = order.UnitPrice,
            BasePrice = order.BasePrice,
            DiscountRate = order.DiscountRate,
            DiscountAmount = order.DiscountAmount,
            SubtotalPrice = order.SubtotalPrice,
            TaxRate = order.TaxRate,
            TaxAmount = order.TaxAmount,
            TotalPrice = order.TotalPrice,
            Status = order.Status,
            Country = order.Country,
            AppliedCouponCode = order.AppliedCouponCode
        };
    }

    private static string GenerateOrderNumber()
    {
        return $"ORD-{Guid.NewGuid().ToString().ToUpper()}";
    }
}
